arches_for_science/trigger_functions.py

```diff
@@ -0,0 +1,184 @@
+CALCULATE_MULTICARD_PRIMARY_DESCRIPTOR_SINGLE = r"""
+DECLARE
+    fn_config jsonb;
+    name_template text;
+    map_popup_template text;
+    description_template text;
+    all_templates text[];
+    this_template text;
+    loop_index int;
+    descriptor_key text;
+
+    graph UUID;
+    resourceid UUID;
+    node_alias text;
+    alias_with_separators text;
+    node_for_alias UUID;
+    nodegroup_for_alias UUID;
+    localized_string_node_value jsonb;
+    lang text;
+    inner18n_obj_for_lang jsonb;
+    resolved_node_value_for_lang text;
+    working_string text;
+    localized_result_all jsonb;
+    localized_result_name_only jsonb;
+
+BEGIN
+
+SELECT NEW.resourceinstanceid INTO resourceid;
+SELECT graphid INTO graph FROM resource_instances WHERE resourceinstanceid = resourceid;
+
+SELECT config
+INTO fn_config
+FROM functions_x_graphs
+WHERE
+    functionid = '00b2d15a-fda0-4578-b79a-784e4138664b'
+    AND functions_x_graphs.config IS NOT NULL
+    AND graph = graphid;
+
+IF FOUND THEN
+    SELECT JSONB_OBJECT('{}') INTO localized_result_all;
+    SELECT JSONB_OBJECT('{}') INTO localized_result_name_only;
+    SELECT fn_config -> 'descriptor_types' -> 'name' -> 'string_template' INTO name_template;
+    SELECT fn_config -> 'descriptor_types' -> 'map_popup' -> 'string_template' INTO map_popup_template;
+    SELECT fn_config -> 'descriptor_types' -> 'description' -> 'string_template' INTO description_template;
+
+    SELECT ARRAY[name_template, map_popup_template, description_template] INTO all_templates;
+    SELECT 0 INTO loop_index;
+    FOR this_template IN SELECT UNNEST(all_templates)
```

Comment:
Overall, the approach is:
```
For each descriptor...
    For each node alias in the template...
        For each language in the i18n obj for that node...
            Update the template string to replace that specific node
```
I'm doing string replacements, as I go, e.g.
`Prefix <value1> Middle Value <value2> Suffix`
becomes
`Prefix ResolvedValue1 Middle Value <value2> Suffix`
and then
`Prefix ResolvedValue1 Middle Value ResolvedValue2 Suffix`
.

This involves a good deal of casting text (on which I did the string replacement) to JSON (so I can store the intermediate results). It just seemed to me that if I created a variable to hold the all the localized strings _as strings (and not JSON)_, then I'd be creating a nested tree very similar (identical) to the one I could just update as I go, and this would make the code harder to read. I'm open to ideas here.